api/src/wfl/sink.clj

```diff
@@ -242,21 +243,105 @@
+               ORDER BY id LIMIT 1"]
+    (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+      (postgres/throw-unless-table-exists tx details)
+      (first (jdbc/query tx (format query details))))))
+
+(defn ^:private pop-job-queue!
+  [{:keys [details] :as _sink} {:keys [id] :as _job}]
+  {:pre [(some? id) (integer? id)]}
+  (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+    (jdbc/update! tx details {:consumed (utc-now)} ["id = ?" id])))
+
+(defn ^:private update-datarepo-job-statuses
+  "Fetch and record the status of all 'running' jobs from tdr created by the
+   `sink`. Throws `UserException` when any new job status is `failed`."
+  [{:keys [details] :as _sink}]
+  (letfn [(read-active-jobs []
+            (let [query "SELECT id, job, workflow FROM %s
+                         WHERE status = 'running'
+                         ORDER BY id ASC"]
+              (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+                (map (juxt :id :job :workflow)
+                     (jdbc/query tx (format query details))))))
+          (write-job-status [job-id status]
+            (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+              (jdbc/update! tx details
+                            {:status status :updated (utc-now)}
+                            ["id = ?" job-id])))]
+    (doseq [[id job-id workflow] (read-active-jobs)]
+      (let [job    (datarepo/job-metadata job-id)
+            status (-> job :job_status str/lower-case)]
+        (write-job-status id status)
+        (when (= "failed" status)
+          (throw (UserException. "A DataRepo ingest job failed"
+                                 {:job      job
+                                  :workflow workflow})))))))
+
+(defn ^:private to-dataset-row
+  "Use `fromOutputs` to coerce the `workflow` outputs into a row in the dataset
+   where `fromOutputs` describes a mapping from workflow outputs to columns in
+   the dataset table."
+  [fromOutputs {:keys [outputs] :as workflow}]
+  (when-not (map? fromOutputs)
+    (throw (IllegalStateException. "fromOutputs is malformed")))
+  (try
+    (rename-gather outputs fromOutputs)
+    (catch Exception cause
+      (throw (ex-info "Failed to coerce workflow outputs to dataset columns"
+                      {:fromOutputs fromOutputs :workflow workflow}
+                      cause)))))
+
+(defn ^:private start-ingesting-outputs
+  "Start ingesting the `workflow` outputs as a new row in the target dataset
+   and push the tdr ingest job into the `sink`'s job queue."
+  [{:keys [dataset table fromOutputs] :as sink}
+   {:keys [uuid] :as workflow}]
+  (let [file (storage/gs-url "broad-gotc-dev-wfl-ptc-test-outputs" (str uuid ".json"))]
+    (-> (to-dataset-row fromOutputs workflow)
+        (json/write-str :escape-slash false)
+        (storage/upload-content file))
+    (->> (datarepo/ingest-table (:id dataset) file table)
+         (assoc {:status "running" :workflow uuid} :job)
```

Comment:
We're positive that if we've gotten past `datarepo/ingest-table` successfully, then "running" is the correct status for the job?  Can we get this from an external service instead of hard-coding?  I noticed that `datarepo/ingest-table` pulls out the job id from the response and returns that, I wonder if the status is available in the response.

Reply:
I would prefer to get them form another service alas no service enumerates them. These values are lifted from the prod swagger page so I'm sure they're correct for now...

Reply:
It is available in the response yes but we know it's running when we create the job. I suppose we should use the value tdr gives us but that's a bigger change.

Reply:
Agreed, that's a bigger scope.